### ISSUE TITLE
Inconsistency in Actor Net input dimension vs Morphology

### ISSUE THREAD

**Issue by TaiE64** [arclab-hku/bipedal_locomotion_for_quadrupedal_robots#2]
Hi, I noticed a contradiction regarding the Actor Net input in the paper:  
Text (Page 10): The Actor Net section states the input $\hat{e}_t$ is $\in \mathbb{R}^6$ (only estimated velocity + CoM-CoP), explicitly excluding morphology.  
Equation (14) & Figure 8: These define the estimator output to include morphology ($\hat{o}_t^{mor}$), and the "Generalization" section (Page 4) implies the controller relies on this for adaptability.  
Question:Does the Actor Net actually receive the estimated morphology as input?   
Thanks!

**Comment by ErdongXiao** [comment]
Thank you very much for pointing this out and for your question.

The morphology identification module is primarily designed to enable the training of a unified reinforcement learning (RL) controller that generalizes across multiple robot morphologies in the training stage. Further details on this technique can be found in our previous work: https://ieeexplore.ieee.org/document/10463132/.

However, it is not necessarily required in the baseline TumblerNet framework, which targets deployment on a specific robot morphology (e.g., Go-1 or Lite-3 only).

In practice, if you do not intend to train a general policy that could be directly deployed on multiple robots. It is feasible to either keep the input obtained from the estimator 6 dimensions (as you mentioned, base velocity + CoM-CoP), or fix the last N dimensions as constant [0,0, ...,0]^T. 

If you simply needed to deploy the policy, which has already been trained within only one morphology of robot, on the other morphologies of robots, there are also some alternative methods to resolve this. 

For instance, implementing the Imitation Learning method is considered an efficient solution, as also illustrated in some of our other works: https://advanced.onlinelibrary.wiley.com/doi/full/10.1002/adrr.202500036 and https://www.nature.com/articles/s44182-025-00048-x

**Comment by TaiE64** [comment]
Thank you very much for your response:D
I am a third-year undergraduate student majoring in Robotics and AI at UCL. I am currently attempting to reproduce your work and have successfully trained the policy in Isaac Lab, achieving the transition from a quadrupedal stance to bipedal standing and walking. 
However, I am facing two major challenges: the robot's forelimbs exhibit significant jitter during the bipedal phase, and my attempts at Sim2Sim transfer to MuJoCo have consistently failed. Could you kindly offer any advice or insights on how to mitigate these dynamics mismatches?
( I did not include the Actuator Net  because I do not have the real-world actuator data required for training it.)

**Comment by ErdongXiao** [comment]
> Thank you very much for your response:D I am a third-year undergraduate student majoring in Robotics and AI at UCL. I am currently attempting to reproduce your work and have successfully trained the policy in Isaac Lab, achieving the transition from a quadrupedal stance to bipedal standing and walking. However, I am facing two major challenges: the robot's forelimbs exhibit significant jitter during the bipedal phase, and my attempts at Sim2Sim transfer to MuJoCo have consistently failed. Could you kindly offer any advice or insights on how to mitigate these dynamics mismatches? ( I did not include the Actuator Net because I do not have the real-world actuator data required for training it.)

Sure, I’m glad that you’re interested in this task, and I’m more than happy to give you some of my personal suggestions.

Regarding your first issue, I think the phenomenon you describe reveals that your policy needs additional regularization on the forelimbs’ motion, as its output appears to be some noisy, trembling motion. A suggested method to resolve this is to add some soft constraints during training or fine-tuning. You could try designing a reward/penalty function (for instance, L2 distance to the target) to either constrain the end-effector of the forelimb to a specific position in the body frame, or to a plane of a specific height in the world frame (this would be feasible for the flat-plane locomotion task).

Regarding your second issue, which is about the Sim2Sim gap, this is very common. I do suggest that the most efficient way to minimize the dynamics mismatch is by using the actuator net (given the joint position, velocity, and target joint position over a few time steps, it outputs the actual torque) collected from the target simulation environment to fine-tune or train your policy. At the same time, assign a larger range to Domain Randomization of the time delay (latency). This should be similar to some common method that minimizing the Sim2Real gap. Before doing this, you could print logs of the target joint position, target joint torque, and the actual joint torque applied in both simulators under the same command, and compare the differences.

Last but not least, I also suggest you double-check the Kp, Kd, and frequencies between your training and deployment environments. Normally, there are 3 different kinds of frequencies. 1. The policy network’s inference frequency is suggested 50 Hz, which means the actor receives new input and produces a new output every 0.02 seconds (please make sure this is the same on both platforms). 2. The command sending frequency, which can be higher than 200 Hz. It  means you repeat the action command and send it to the robot in your deployment environment at least every 0.005 seconds. 3. The simulation updating frequency,  the time step length that you update the state of every items in your simulation world (Mujoco) should ideally be even smaller than this value. (in real world you do not need to concern about this.)

If you’d like to provide further details about the problem you’re facing, we are also very happy to have more discussion on it or provide further help, please feel free to email me at exiao@connect.hku.hk.
